pennylane/workflow/return_types_spec.rst

```diff
@@ -0,0 +1,201 @@
+
+.. _ReturnTypeSpec:
+
+Return Type Specification
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This page describes the shape and type of the numerical output from executing a quantum circuit
+in PennyLane.
+
+The specification applies for the entire workflow, from the device instance all the
+way up to the ``QNode``.  The result object corresponding to a given circuit
+will match whether the circuit is being passed to a device, processed
+by a transform, having it's derivative bound to an ML interface, or returned from a ``QNode``.
+
+While this document says ``tuple`` and includes examples using ``tuple`` throughout this document, the
+return type specification allows ``tuple`` and ``list`` to be used interchangably.
+When examining and postprocessing
+results, you should always allow for a ``list`` to be substituted for a ``tuple``. Given their
+improved performance and protection against unintended side-effects, ``tuple``'s are recommended
+over ``list`` where feasible.
+
+The level of priority for dimensions from outer dimension to inner dimension is:
+
+1. Quantum Script in batch. No squeezing
+2. Shot choice in a shot vector. Squeezed if no shot vector
+3. Measurement in the quantum script. Squeezed out if only one measurement
+4. Parameter broadcasting. Squeezed out if no parameter-broadcasting.  Adds to array shape instead of adding tuple nesting.
+5. Fundamental measurement shape.
+
+Individual measurements
+-----------------------
+
+Each individual measurement corresponds to its own type of result. This result can be
+a Tensor-like (Python number, numpy array, ML array), but may also be any other type of object.
+For example, :class:`~.CountsMP` corresponds to a dictionary. We can also imagine a scenario where
+a measurement corresponds to some other type of custom data structure.
+
+The result corresponding to a given measurement process should also be independent of any other
+measurements present at the same time.  For example, requesting the probability ``qml.probs(wires=0)``
+should not affect the result corresponding to an expectation value ``qml.expval(qml.Z(0))``.
```

Comment:
What does this mean? Don't we expect probs and expvals to potentially be correlated (e.g., finite shots, both in the same basis)?

Reply:
Just deleting this paragraph. Reaslizing it probably confused more than it clarifies.